src/chain_access.rs

```diff
@@ -1,27 +1,284 @@
+    ) -> Self {
         Self {
-            esplora_client,
+            esplora: esplora_client,
             filter,
+            watched_txs: HashSet::new(),
+            watched_outputs: HashSet::new(),
+            synced_tip,
         }
     }
 
-    pub fn sync(&self, _confirm: &(dyn Confirm + Sync)) -> Result<(), RuntimeError> {
-        // TODO: sync with the chain
+    pub(crate) fn sync(&mut self, confirm: &ConfirmWrapper) -> Result<(), Error> {
+        let mut cur_tip = self.esplora.get_tip_hash()?;
+
+        while self.synced_tip != cur_tip {
+            debug!("Currently synced up to block hash: {}", self.synced_tip);
+            debug!(
+                "Current blockchain tip hash: {} -> Start syncing blockchain",
+                cur_tip
+            );
+
+            self.sync_to_tip(confirm, &cur_tip)?;
+            self.synced_tip = cur_tip;
+            cur_tip = self.esplora.get_tip_hash()?;
+        }
+
+        Ok(())
+    }
+
+    fn sync_to_tip(&mut self, confirm: &ConfirmWrapper, cur_tip: &BlockHash) -> Result<(), Error> {
+        self.inform_about_new_block(confirm, cur_tip)?;
+
+        // todo: we need to make sure that the following synchronization is only done up to cur_tip.
+        //       otherwise, we might end up in a situation where a block is being mined while while we're syncing
+        //       which would make corrupt the order of the confirmed transactions as required by the Confirm trait:
+        //       https://github.com/lightningdevkit/rust-lightning/blob/d6321e6e11b3e1b11e26617d3bab0b8c21da0b5b/lightning/src/chain/mod.rs#L128
+
+        loop {
+            self.sync_relevant_txs_for_reorgs(confirm)?;
+
+            let mut confirmed_txs = self.sync_txs()?;
+            confirmed_txs.extend(self.sync_spending_txs()?);
+            debug!("{} confirmed transactions", confirmed_txs.len());
+            debug!("Confirmed transaction list: {:?}", confirmed_txs);
+
+            sort_txs(&mut confirmed_txs);
+
+            for tx in confirmed_txs {
+                confirm.transactions_confirmed(
+                    &tx.block_header,
+                    &[(tx.position, &tx.tx)],
+                    tx.block_height,
+                );
+            }
+
+            match self.filter.drain() {
+                Some(filter_data) => {
+                    filter_data.txs.iter().for_each(|tx| {
+                        self.watched_txs.insert((tx.0, tx.1.clone()));
+                    });
+                    filter_data.outputs.into_iter().for_each(|output| {
+                        let output = WatchedOutputClone {
+                            block_hash: output.block_hash,
+                            outpoint: output.outpoint,
+                            script_pubkey: output.script_pubkey,
+                        };
+                        self.watched_outputs.insert(output);
+                    });
+                }
+                None => break,
+            }
+        }
```

Comment:
Why do we need the `loop`? Couldn't we deal with the data concurrently added to `filter` in the next `sync()` call?

Reply:
Probably we could ... but I think when you call `sync()` you would except that the node is perfectly synced up to the tip after completion

Reply:
Added some clarification here: 3be89bfe9de265f952d7a420e5163c3694ceabdd